posthog/models/entity.py

```diff
@@ -19,6 +19,7 @@ class Entity(PropertyMixin):
     id: Union[int, str]
     type: str
     order: Optional[int]
+    index: int
```

Comment:
This type is wrong and indicates that this doesn't solve the problem.

This may still well be `Null` as there's other cases where Entity objects get created: tests, posthog/api/action.py, posthog/api/utils.py, ExclusionEntities, posthog/models/filters/mixins/retention.py, posthog/models/filters/mixins/sessions.py, posthog/tasks/calculate_cohort.py

All-in-all IMO this removes the comment but the solution is not any better - just more likely to cause bugs down the line.

Reply:
This type is correct.

This solution is better because:
- `index` is deterministic (instead of function call counting magic with a modulo)
- `index` is populated for every call site where it's used, as call sites in the querying engine use processed entities from `EntitiesMixin.entities`
- there is no hidden bug possible because even if later down the line we create a call site where `index` is not populated yet, trying to access it will raise an `AttributeError` instead of returning `None` (that's why the type is non-optional `int`)

Reply:
In practice it's not populated in any of the many callsites I listed in the previous comment. 

Good note re this raising an AttributeError though, still weird.

This is one of those "two strong heads butting" situations, assigning @EDsCODE to mediate since they never go anywhere. :)

Reply:
This typing is pretty peculiar. I think we should use this solution to get rid of the modulo funk but maybe add a comment next to the type that details how it eventually gets populated otherwise later on when we reference this model definition we might be somewhat misled/it might look like a bug.

Reply:
Added a pretty extensive comment :)